src/Symfony/Component/HttpKernel/Controller/ArgumentResolver.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Controller;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\ControllerMetadata\Argument\ArgumentMetadataFactoryInterface;
+
+/**
+ * Responsible for the resolving of arguments passed to an action.
+ *
+ * @author Iltar van der Berg <kjarli@gmail.com>
+ */
+final class ArgumentResolver implements ArgumentResolverInterface
+{
+    /**
+     * @var ArgumentMetadataFactoryInterface
+     */
+    private $argumentMetadataFactory;
+
+    /**
+     * @var ArgumentValueResolverInterface[]
+     */
+    private $argumentValueResolvers;
+
+    /**
+     * @param ArgumentMetadataFactoryInterface $argumentMetadataFactory
+     * @param ArgumentValueResolverInterface[] $argumentValueResolvers
+     */
+    public function __construct(ArgumentMetadataFactoryInterface $argumentMetadataFactory = null, array $argumentValueResolvers = [])
+    {
+        $this->argumentMetadataFactory = $argumentMetadataFactory;
+        $this->argumentValueResolvers = $argumentValueResolvers;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getArguments(Request $request, $controller)
+    {
+        $arguments = array();
+
+        foreach ($this->argumentMetadataFactory->createArgumentMetadata($controller) as $metadata) {
+            $isResolved = false;
+            foreach ($this->argumentValueResolvers as $resolver) {
```

Comment:
That would cause an issue with `(Request $request, $foo)` where foo wouldn't exist. Because at that point it would return true while the argument couldn't be found.

I've also thought about the guard clause you've posted but this case will actually cause 2 arguments to be resolved if they overlap. That means if `(Request $request)` is given and 2 resolvers support this:
 - Request as class
 - 'request' in the attribute bag for example
this would mean you'd call the method with 2 parameters instead of 1.